README.md

```diff
@@ -1,9 +1,241 @@
+                self.postings[dim_id] = PostingList::new_one(id, weight);
+            }
+        }
+    }
+    // given that there are no holes in the internal ids and that we are not deleting from the index
+    // we can just use the id as a proxy the count
+    self.vector_count = max(self.vector_count, id as usize);
+}
+```
+
+Having both encoders should help us build a more robust search mechanism, that can handle both the natural language and 
+code-specific queries.
+
+### Search process
+
+The search process is quite straightforward. The user input is passed to both encoders, and the resulting vectors are
+used to query both Qdrant collections at the same time. The results are then merged with duplicates removed and returned 
+back to the user. 
+
+## Architecture
+
+The demo uses [FastAPI](https://fastapi.tiangolo.com/) framework for the backend and [React](https://reactjs.org/) for 
+the frontend layer. 
+
+![Architecture of the code search demo](images/architecture-diagram.png)
+
+The demo consists of the following components:
+- [React frontend](/frontend) - a web application that allows the user to search over Qdrant codebase
+- [FastAPI backend](/code_search/service.py) - a backend that communicates with Qdrant and exposes a REST API
+- [Qdrant](https://qdrant.tech/) - a vector search engine that stores the data and performs the search
+- Two neural encoders - one trained on the natural language and one for the code-specific tasks
+
+There is also an additional indexing component that has to be run periodically to keep the index up to date. It is also
+part of the demo, but it is not directly exposed to the user. All the required scripts are documented below, and you can 
+find them in the [`tools`](/tools) directory.
+
+The demo is, as always, open source, so feel free to check the code in this repository to see how it is implemented.
+
+## Usage
+
+As every other semantic search system, the demo requires a few steps to be set up. First of all, the data has to be
+ingested, so we can then use the created index for our queries.
+
+### Data indexing
+
+Qdrant is used as a search engine, so you will need to have it running somewhere. You can either use the local container
+or the Cloud version. If you want to use the local version, you can start it with the following command:
+
+```shell
+docker run -p 6333:6333 -p 6334:6334 \
+    -v $(pwd)/qdrant_storage:/qdrant/storage:z \
+    qdrant/qdrant
+```
+
+However, the easiest way to start using Qdrant is to use our Cloud version. You can sign up for a free tier 1GB cluster 
+at [https://cloud.qdrant.io/](https://cloud.qdrant.io/).
+
+Once the environment is set up, you can configure the Qdrant instance and build the index by running the following 
+commands:
+
+```shell
+export QDRANT_URL="http://localhost:6333"
+
+# For the Cloud service you need to specify the api key as well
+# export QDRANT_API_KEY="your-api-key"
+
+bash tools/download_and_index.sh
```

Comment:
You could `chmod u+x tools/download_and_index.sh`
- Then you wouldn't need to preface with `bash` 

When I run the script, I run into these issues:

- "tools/index_qdrant.sh: line 14: python: command not found"
  - I've installed Python 3.10, which includes `python3` as an executable
  - While I addressed the problem with a soft link (`ln -s /usr/bin/python3 /usr/local/bin/python`), I'm not sure everyone would catch that error
  
I then run the `download_and_index.sh` script again, and I get this error:

`error: 'rust-analyzer' is not installed for the toolchain 'stable-x86_64-unknown-linux-gnu'`

I discuss this more in an earlier comment.